src/Chainweb/Version/Utils.hs

```diff
@@ -0,0 +1,434 @@
+{-# INLINE someChainId #-}
+
+-- | Sometimes, in particular for testing and examples, some fixed chain id is
+-- needed, but it doesn't matter which one. This function provides some valid
+-- chain ids for the chainweb at the given height.
+--
+someChainIdAt :: HasCallStack => HasChainwebVersion v => v -> BlockHeight -> ChainId
+someChainIdAt v h = head . toList $ chainIdsAt v h
+    -- 'head' is guaranteed to succeed because the empty graph isn't a valid chain
+    -- graph.
+{-# INLINE someChainIdAt #-}
+
+isGraphChange :: HasChainwebVersion v => v -> BlockHeight -> Bool
+isGraphChange v h = M.member h (chainGraphs v)
+{-# INLINE isGraphChange #-}
+
+-- -------------------------------------------------------------------------- --
+-- Block Count
+
+-- | The numbers of blocks on a chain up to the given height
+--
+-- Precondition: h > genesisHeight
+--
+blockCountAt
+    :: HasCallStack
+    => HasChainwebVersion v
+    => HasChainId cid
+    => v
+    -> cid
+    -> BlockHeight
+    -> Natural
+blockCountAt v cid h
+    | h < gh = error $ "Chainweb.Version.Utils.blockCount: requested block height " <> sshow h
+        <> " is smaller than the genesis height " <> sshow gh <> "."
+    | otherwise = 1 + int h - int gh
+  where
+    gh = genesisHeight (_chainwebVersion v) (_chainId cid)
+
+-- | The block count accross all chains at a given block height
+--
+globalBlockCountAt
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> Natural
+globalBlockCountAt v h = sum
+    $ fmap (\c -> blockCountAt v c h)
+    $ toList
+    $ chainIdsAt v h
+
+globalBlockRateAt
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> Double
+globalBlockRateAt v h = int r / int (chainCountAt v h)
+  where
+    BlockRate r = blockRate (_chainwebVersion v)
+
+-- -------------------------------------------------------------------------- --
+-- Cut Heights
+
+-- | The the average over all possible cuts for which at least one chain has the
+-- given block height.
+--
+-- Note, that the result isn't accurate for block heights around a chain graph
+-- change.
+--
+avgCutHeightAt :: HasChainwebVersion v => v -> BlockHeight -> CutHeight
+avgCutHeightAt v h = int $ int h * chainCountAt v h
+{-# INLINE avgCutHeightAt #-}
+
+-- | Cut height intervals for the chain graphs of a chainweb version
+--
+-- Post-condition:
+--
+-- @
+-- 0 == minimum $ M.keys $ cutHeights v
+-- @
+--
+chainGraphsByCutHeight
+    :: HasChainwebVersion v
+    => v
+    -> M.Map CutHeight ChainGraph
+chainGraphsByCutHeight = M.fromAscList
+    . fmap (\(h,g) -> (int h * int (order g), g))
+    . M.toAscList
+    . chainGraphs
+{-# INLINE chainGraphsByCutHeight #-}
+
+-- chainGraphAtCutHeight = atCutHeight h
+
+-- | The chain graph at the given cut height
+--
+-- Note, that the result isn't accurate during a chain graph change
+--
+chainGraphAtCutHeight :: HasChainwebVersion v => v -> CutHeight -> ChainGraph
```

Comment:
Do we need all the versions of everything for both CutHeight and BlockHeight? Would suggest eliding unused stuff

Reply:
Probably not all of it. However, I wrote this module after going through several cycles of adding, deleting, and rewriting utils while working downstream in the code base. So, I made this more exhaustive to note have to go back and forth.

I propose that we keep it until the graph-transition project finished and come back and clean it up when we are done with it. I think, that may also be a good time to revisit the decision of using `Map` (and `List.NonEmpty` in `Chainweb.Version`).